src/s3_payload.rs

```diff
@@ -0,0 +1,151 @@
+use crate::exports::edgee::protocols::data_collection::Dict;
+use anyhow::anyhow;
+use aws_credential_types::Credentials;
+use aws_sigv4::http_request::{
+    sign, PayloadChecksumKind, SignableBody, SignableRequest, SigningParams, SigningSettings,
+};
+use aws_sigv4::sign::v4;
+use aws_smithy_runtime_api::client::identity::Identity;
+use chrono::offset::Utc;
+use chrono::DateTime;
+use std::collections::HashMap;
+use std::time::SystemTime;
+use uuid::Uuid;
+
+pub struct S3Config {
+    pub access_key: String,
+    pub secret_key: String,
+    pub session_token: String, // could be empty
+    pub region: String,
+    pub bucket: String,
+    pub key_prefix: String, // could be empty
+}
+
+impl S3Config {
+    pub fn new(cred_map: Dict) -> anyhow::Result<Self> {
+        let cred: HashMap<String, String> = cred_map
+            .iter()
+            .map(|(key, value)| (key.to_string(), value.to_string()))
+            .collect();
+
+        let access_key = match cred.get("aws_access_key") {
+            Some(key) => key,
+            None => return Err(anyhow!("Missing AWS Access Key")),
+        }
+        .to_string();
+
+        let secret_key = match cred.get("aws_secret_key") {
+            Some(key) => key,
+            None => return Err(anyhow!("Missing AWS Secret Key")),
+        }
+        .to_string();
```

Comment:
You can use `anyhow::Context` here to simplify the code:
```suggestion
	let secret_key = cred.get("aws_secret_key").context("Missing AWS Secret Key")?;
```

Reply:
nice, thanks 🙏

Reply:
that produces a `&String`, so I'm doing:

```rust
let access_key = cred.get("aws_access_key").context("Missing AWS Access Key")?.to_string();
```

is there a better way?